modules/ding_webtrekk/ding_webtrekk.module

```diff
@@ -90,35 +153,170 @@ function ding_webtrekk_views_post_execute(&$view) {
+      'p_mat_indexno' => $entity->getClassification(),
+      'p_mat_lang' => $entity->getLanguage(),
+      'p_mat_source' => $entity->getAc_source(),
+      'p_mat_category' => implode(';', $entity->reply->getGenre()),
+    ];
+    ding_webtrekk_add_page_parameters($params);
+  }
+
+  // Look for at reserve button on this ding entity and setup a Webtrekk event
+  // if present. These could be added in every view_mode and page.
+  if (isset($entity->content['ding_entity_buttons'])) {
+    foreach (element_children($entity->content['ding_entity_buttons']) as $delta) {
+      foreach ($entity->content['ding_entity_buttons'][$delta] as $key => &$button) {
+        // Reserve button.
+        if (isset($button['#path']) && $button['#path'] == $entity_path . '/reserve') {
+          $button['#options']['attributes']['class'][] = 'ding-webtrekk-event';
+          $event_data = json_encode([
+            'linkId' => 'Reserver',
+            'customClickParameter' => [50 => $entity->getId()],
+          ]);
+          $button['#options']['attributes']['data-ding-webtrekk-event'] = $event_data;
+        }
+        // See online button.
+        elseif (isset($button['#attributes']) && in_array('button-see-online', $button['#attributes']['class'])) {
+          $button['#attributes']['class'][] = 'ding-webtrekk-event';
+          $event_data = json_encode([
+            'linkId' => 'Se online',
+            'customClickParameter' => [51 => $entity->getId()],
+          ]);
+          $button['#attributes']['data-ding-webtrekk-event'] = $event_data;
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Implements hook_preprocess_ding_entity_rating_display().
+ */
+function ding_webtrekk_preprocess_ding_entity_rating_display(&$variables) {
+  $variables['classes_array'][] = 'ding-webtrekk-rating-event';
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
  *
- * @param array $value_map
- *   Key, value collection of parameters to be considered.
+ * Attached Webtrekk events to the renew loan buttons.
  */
-function _ding_webtrekk_set_data_layer_values(array $value_map) {
-  $hook = 'ding_webtrekk_set_data_layer_values';
-  $modules = module_implements($hook);
-  foreach ($modules as $module) {
-    if (function_exists($module . '_' . $hook)) {
-      $params[] = module_invoke($module, $hook);
+function ding_webtrekk_form_ding_loan_loans_form_alter(&$form, &$form_state) {
+  // We'll only attach the renew events if there's loans that can be renewed.
+  // Ding loan form checks this and only sets this element if that's the case.
+  if (!isset($form['actions_container']['select_all'])) {
+    return;
+  }
+
+  // Ding loan stores the loans in the this form element and we take advantage
+  // of that to get the material ids for the renewable loans.
+  $loans = $form['items']['#value'];
+  $material_ids = [];
+  foreach ($loans as $loan) {
+    if ($loan->renewable) {
+      $material_ids[$loan->id] = $loan->getEntity()->getId();
```

Comment:
Do we really want/need to track every element renewed on the website? That seems a bit execessive 🤔 .

Reply:
I agree, but this was the requirement.